tieredstore/src/main/java/org/apache/rocketmq/store/tiered/container/TieredFileSegment.java

```diff
@@ -0,0 +1,529 @@
+        int finalBufferSize = bufferSize;
+        try {
+            inflightCommitRequest = commit0(inputStream, commitPosition, bufferSize, fileType != FileSegmentType.INDEX)
+                .thenApply(result -> {
+                    if (result) {
+                        if (fileType == FileSegmentType.COMMIT_LOG && bufferList.size() > 0) {
+                            commitMsgQueueOffset = MessageBufferUtil.getQueueOffset(bufferList.get(bufferList.size() - 1));
+                        }
+                        commitPosition += finalBufferSize;
+                        return true;
+                    }
+                    sendBackBuffer(inputStream);
+                    return false;
+                }).exceptionally(e -> handleCommitException(inputStream, e))
+                .whenComplete((result, e) -> {
+                    if (commitLock.availablePermits() == 0) {
+                        LOGGER.debug("TieredFileSegment#commitAsync: commit cost: {}ms, file: {}, item count: {}, buffer size: {}", stopwatch.elapsed(TimeUnit.MILLISECONDS), getPath(), bufferList.size(), finalBufferSize);
+                        commitLock.release();
+                    } else {
+                        LOGGER.error("[Bug]TieredFileSegment#commitAsync: commit lock is already released: available permits: {}", commitLock.availablePermits());
+                    }
+                });
+            return inflightCommitRequest;
+        } catch (Exception e) {
+            handleCommitException(inputStream, e);
+            if (commitLock.availablePermits() == 0) {
+                LOGGER.debug("TieredFileSegment#commitAsync: commit cost: {}ms, file: {}, item count: {}, buffer size: {}", stopwatch.elapsed(TimeUnit.MILLISECONDS), getPath(), bufferList.size(), finalBufferSize);
+                commitLock.release();
+            } else {
+                LOGGER.error("[Bug]TieredFileSegment#commitAsync: commit lock is already released: available permits: {}", commitLock.availablePermits());
+            }
+        }
+        return CompletableFuture.completedFuture(false);
+    }
+
+    private boolean handleCommitException(TieredFileSegmentInputStream inputStream, Throwable e) {
+        Throwable cause = e.getCause() != null ? e.getCause() : e;
+        sendBackBuffer(inputStream);
+        long realSize = 0;
+        if (cause instanceof TieredStoreException && ((TieredStoreException) cause).getPosition() > 0) {
+            realSize = ((TieredStoreException) cause).getPosition();
+        }
+        if (realSize <= 0) {
+            realSize = getSize();
+        }
+        if (realSize > 0 && realSize > commitPosition) {
+            LOGGER.error("TieredFileSegment#handleCommitException: commit failed: file: {}, try to fix position: origin: {}, real: {}", getPath(), commitPosition, realSize, cause);
+            // TODO check if this diff part is uploaded to OSS
```

Comment:
OSS? tiered-storage should not be limited to the specific implementation.

Reply:
The `check` here could be an independent method or we could implement it in `#commit0`.